compressors/rANS.py

```diff
@@ -0,0 +1,335 @@
+"""Streaming rANS (range Asymmetric Numeral Systems) implementation
+
+## Core idea
+- the theoretical rANS Encoder maintains an integer `state` 
+- For each symbol s, the state is updated by calling: 
+    ```python
+    # encode step
+    state = rans_base_encode_step(s, state)
+    ```
+    the decoder does the reverse by decoding the s and retrieving the prev state
+    ```python
+    # decode step
+    s, state = rans_base_decode_step(state)
+    ```
+- In the theoretical rANS version, the state keeps on increasing every time we call `rans_base_encode_step`
+  To make this practical, the rANS encoder ensures that after each encode step, the `state` lies in the acceptable range
+  `[L, H]`, where `L,H` are predefined interval values.
+
+  ```
+  state lies in [L, H], after every encode step
+  ```
+  To ensure this happens the encoder shrinks the `state` by streaming out its lower bits, *before* encoding each symbol. 
+  This logic is implemented in the function `shrink_state`. Thus, the full encoding step for one symbol is as follows:
+
+  ```python
+    ## Encoding one symbol
+    # output bits to the stream to bring the state in the range for the next encoding
+        state, out_bits = self.shrink_state(state, s)
+        encoded_bitarray = out_bits + encoded_bitarray
+
+    # core encoding step
+    state = self.rans_base_encode_step(s, state)
+  ```
+
+  The decoder does the reverse operation of `expand_state` where, after decoding a symbol, it reads in the a few bits to
+  re-map and expand the state to lie within the acceptable range [L, H]
+  Note that `shrink_state` and `expand_state` are inverses of each other
+
+  Thus, the  full decoding step 
+  ```python
+    # base rANS decoding step
+    s, state = self.rans_base_decode_step(state)
+
+    # remap the state into the acceptable range
+    state, num_bits_used_by_expand_state = self.expand_state(state, encoded_bitarray)
+  ```
+- For completeness: the acceptable range `[L, H]` are given by:
+  L = RANGE_FACTOR*total_freq
+  H = (2**NUM_BITS_OUT)*RANGE_FACTOR*total_freq - 1)
+  Why specifically these values? Look at the *Streaming-rANS* section on https://kedartatwawadi.github.io/post--ANS/
+
+
+## References
+1. Original Asymmetric Numeral Systems paper:  https://arxiv.org/abs/0902.0271
+2. Interactive write-up on rANS: https://kedartatwawadi.github.io/post--ANS/
+
+## TODO
+1. Implement alias-coding style rANS
```

Comment:
what's this (in short)?

Reply:
I will briefly mention this tomorrow.. but as a hint look at the puzzle I posted on slack!